src/Dfc.CourseDirectory.WebV2/Features/Providers/EditProviderType.cs

```diff
@@ -40,25 +52,77 @@ public async Task<Command> Handle(Query request, CancellationToken cancellationT
+
             return new Command()
             {
                 ProviderId = provider.Id,
-                ProviderType = provider.ProviderType
+                ProviderType = provider.ProviderType,
+                ProviderTLevelDefinitions = ProviderTLevelDefinitions(tLevelDefinitions, providerTLevelDefinitions.Select(pd => pd.TLevelDefinitionId))
             };
         }
 
-        public async Task<Success> Handle(Command request, CancellationToken cancellationToken)
+        public async Task<OneOf<ModelWithErrors<Command>, Success>> Handle(Command request, CancellationToken cancellationToken)
         {
+            var tLevelDefinitions = await _sqlQueryDispatcher.ExecuteQuery(new SqlQueries.GetTLevelDefinitions());
+
+            var validationResult = await new CommandValidator(tLevelDefinitions).ValidateAsync(request);
+
+            if (!validationResult.IsValid)
+            {
+                request.ProviderTLevelDefinitions = ProviderTLevelDefinitions(tLevelDefinitions, request.SelectedProviderTLevelDefinitionIds);
+                return new ModelWithErrors<Command>(request, validationResult);
+            }
+
             await _cosmosDbQueryDispatcher.ExecuteQuery(new UpdateProviderType()
             {
                 ProviderId = request.ProviderId,
                 ProviderType = request.ProviderType
             });
 
+            await _sqlQueryDispatcher.ExecuteQuery(new SqlQueries.UpsertProviderTLevelDefinitions
+            {
+                ProviderId = request.ProviderId,
+                TLevelDefinitionIds = request.ProviderType.HasFlag(ProviderType.TLevels)
+                    ? request.SelectedProviderTLevelDefinitionIds ?? Enumerable.Empty<Guid>()
+                    : Enumerable.Empty<Guid>()
+            });
+
             // Remove this provider from the cache - subsequent requests will re-fetch updated record
             await _providerInfoCache.Remove(request.ProviderId);
 
             return new Success();
         }
+
+        private class CommandValidator : AbstractValidator<Command>
+        {
+            public CommandValidator(IEnumerable<SqlModels.TLevelDefinition> tLevelDefinitions)
+            {
+                When(c => c.ProviderType.HasFlag(ProviderType.TLevels), () =>
+                {
+                    RuleFor(c => c.SelectedProviderTLevelDefinitionIds)
+                        .NotEmpty()
+                        .WithMessage("Select the T Levels this provider can offer");
+
+                    When(c => c.SelectedProviderTLevelDefinitionIds?.Any() ?? false, () =>
+                    {
+                        RuleFor(c => c.SelectedProviderTLevelDefinitionIds)
+                            .Must(ids => ids.All(id => tLevelDefinitions.Any(d => d.TLevelDefinitionId == id)));
+                    });
+                });
+            }
+        }
+
+        private static IEnumerable<ProviderTLevelDefinitionViewModel> ProviderTLevelDefinitions(
```

Comment:
```suggestion
        private static IEnumerable<ProviderTLevelDefinitionViewModel> GetProviderTLevelDefinitions(
```

Reply:
Any particular reason?

Reply:
Method names should be commands, IMO. This reads like a property.